### PR TITLE
8327461: KeyStore getEntry is not thread-safe

### PR DESCRIPTION
**Pull request by haimaychao**
Change was made to engineGetEntry() in PKCS12KeyStore to extract the key and certificate chain from Entry only once. This is because the entry may get updated between engineGetKey() and engineGetCertificateChain() which causes inconsistent result. A new test was added to assess and manipulate PKCS12KeyStore with read and write operations concurrently from multiple threads. Thanks!

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8327461](https://bugs.openjdk.org/browse/JDK-8327461): KeyStore getEntry is not thread-safe (**Bug** - P4)

### Reviewers
 * [Daniel Jeliński](https://openjdk.org/census#djelinski) (@djelinski - **Reviewer**) ⚠️ Review applies to [cbea5228](https://git.openjdk.org/jdk/pull/18156/files/cbea5228d1ec4c4e7948d9a4e8c6dc7e59b1b037)
 * [Weijun Wang](https://openjdk.org/census#weijun) (@wangweij - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18156/head:pull/18156` \
`$ git checkout pull/18156`

Update a local copy of the PR: \
`$ git checkout pull/18156` \
`$ git pull https://git.openjdk.org/jdk.git pull/18156/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18156`

View PR using the GUI difftool: \
`$ git pr show -t 18156`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18156.diff">https://git.openjdk.org/jdk/pull/18156.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18156#issuecomment-1984032806)

### CODE REVIEW

File: src/java.base/share/classes/sun/security/pkcs12/PKCS12KeyStore.java [hunk 1]
```diff
@@ -1333,9 +1343,9 @@ public KeyStore.Entry engineGetEntry(String alias,
                         (KeyStore.PasswordProtection)protParam;
                 char[] password = pp.getPassword();
 
-                Key key = engineGetKey(alias, password);
+                Key key = internalGetKey(entry, password);
```
[the remark this answers is not in the document]
Comment (haimaychao) [reply src/java.base/share/classes/sun/security/pkcs12/PKCS12KeyStore.java:1346]: Fixed.

**Comment by johnshajiang** [comment]
Should method `engineEntryInstanceOf` be also concerned?

**Comment by haimaychao** [comment]
engineEntryInstanceOf has only one direct access to the entries map. After obtaining the entry object, its type is checked against the specified entryClass, which do not involve accessing the entries map again.

File: src/java.base/share/classes/sun/security/pkcs12/PKCS12KeyStore.java [hunk 2]
```diff
@@ -1306,7 +1326,7 @@ public KeyStore.Entry engineGetEntry(String alias,
 
         Entry entry = entries.get(alias.toLowerCase(Locale.ENGLISH));
         if (protParam == null) {
-            if (engineIsCertificateEntry(alias)) {
+            if (internalEngineIsCertificateEntry(entry)) {
                 if (entry instanceof CertEntry &&
                     ((CertEntry) entry).trustedKeyUsage != null) {
```
Comment (djelinski) [inline src/java.base/share/classes/sun/security/pkcs12/PKCS12KeyStore.java:1331]: `internalEngineIsCertificateEntry` performs exactly the same checks; you can remove these lines.

File: src/java.base/share/classes/sun/security/pkcs12/PKCS12KeyStore.java [hunk 3]
```diff
@@ -1345,7 +1365,7 @@ public KeyStore.Entry engineGetEntry(String alias,
                     return new KeyStore.SecretKeyEntry((SecretKey)key,
                         entry.attributes);
                 }
-            } else if (!engineIsKeyEntry(alias)) {
+            } else if (!internalEngineIsKeyEntry(entry)) {
```
Comment (djelinski) [inline src/java.base/share/classes/sun/security/pkcs12/PKCS12KeyStore.java:1368]:
```suggestion
            } else {
```

(the matching `if` already checks `internalEngineIsKeyEntry(entry)`)

**Review by djelinski — COMMENTED** [review]
Very nice improvement. Thank you!

File: src/java.base/share/classes/sun/security/pkcs12/PKCS12KeyStore.java (hunk 3 above)
Comment (haimaychao) [reply src/java.base/share/classes/sun/security/pkcs12/PKCS12KeyStore.java]: Removed.

File: src/java.base/share/classes/sun/security/pkcs12/PKCS12KeyStore.java (hunk 2 above)
Comment (haimaychao) [reply src/java.base/share/classes/sun/security/pkcs12/PKCS12KeyStore.java]: Removed.

**Review by djelinski — APPROVED** [review]
LGTM. Thanks!

File: src/java.base/share/classes/sun/security/pkcs12/PKCS12KeyStore.java [hunk 4]
```diff
@@ -295,9 +295,14 @@ static <S> S run(RetryWithZero<S> f, char[] password) throws Exception {
      * (e.g., the given password is wrong).
      */
     public Key engineGetKey(String alias, char[] password)
-        throws NoSuchAlgorithmException, UnrecoverableKeyException
-    {
+        throws NoSuchAlgorithmException, UnrecoverableKeyException {
         Entry entry = entries.get(alias.toLowerCase(Locale.ENGLISH));
+        Key entryKey = internalGetKey(entry, password);
+        return entryKey;
```
[the remark this answers is not in the document]
Comment (haimaychao) [reply src/java.base/share/classes/sun/security/pkcs12/PKCS12KeyStore.java:301]: Done.

File: src/java.base/share/classes/sun/security/pkcs12/PKCS12KeyStore.java [hunk 5]
```diff
@@ -471,15 +476,20 @@ public Key engineGetKey(String alias, char[] password)
      */
     public Certificate[] engineGetCertificateChain(String alias) {
         Entry entry = entries.get(alias.toLowerCase(Locale.ENGLISH));
+        Certificate[] certChain = internalGetCertificateChain(entry);
+        return certChain;
```
[the remark this answers is not in the document]
Comment (haimaychao) [reply src/java.base/share/classes/sun/security/pkcs12/PKCS12KeyStore.java:480]: Done.

File: src/java.base/share/classes/sun/security/pkcs12/PKCS12KeyStore.java [hunk 6]
```diff
@@ -1013,18 +1023,19 @@ public synchronized void engineDeleteEntry(String alias)
             debug.println("Removing entry at alias '" + alias + "'");
         }
 
-        Entry entry = entries.get(alias.toLowerCase(Locale.ENGLISH));
-        if (entry instanceof PrivateKeyEntry keyEntry) {
-            if (keyEntry.chain != null) {
-                certificateCount -= keyEntry.chain.length;
+        Entry entry = entries.remove(alias.toLowerCase(Locale.ENGLISH));
+        if (entry != null) {
```
[the remark this answers is not in the document]
Comment (haimaychao) [reply src/java.base/share/classes/sun/security/pkcs12/PKCS12KeyStore.java:1027]: The API doc states: Returns: the previous value associated with key, or null if there was no mapping for key. It’d be a good practice to check if the entry is not null before proceeding with further operations. Would you please elaborate why it is not needed here?

File: src/java.base/share/classes/sun/security/pkcs12/PKCS12KeyStore.java [hunk 7]
```diff
@@ -1065,6 +1076,10 @@ public int engineSize() {
      */
     public boolean engineIsKeyEntry(String alias) {
         Entry entry = entries.get(alias.toLowerCase(Locale.ENGLISH));
+        return internalEngineIsKeyEntry(entry);
+    }
+
+    private boolean internalEngineIsKeyEntry(Entry entry) {
```
[the remark this answers is not in the document]
Comment (haimaychao) [reply src/java.base/share/classes/sun/security/pkcs12/PKCS12KeyStore.java:1082]: Ok.

**Pull request closed without merging**